File: Scala/src/main/scala/Vehicle.scala
import com.github.nscala_time.time.Imports._

package TollCalculator {

  class Vehicle {
    def uuid = java.util.UUID.randomUUID.toString // Map UUID to customer/registration in billing system

    var lastCharged: Option[DateTime] = None
    var currentDailyFee: SEK = 0

    val setCurrentDailyFee = (amount: SEK) => {
      currentDailyFee = amount
    }
  }

  // Known vehicle types
  case class Car() extends Vehicle
  case class Motorbike() extends Vehicle
  case class Tractor() extends Vehicle
  case class Emergency() extends Vehicle
  case class Diplomat() extends Vehicle
  case class Foreign() extends Vehicle
  case class Military() extends Vehicle
  // Unknown vehicle type
  case class UnknownVehicle() extends Vehicle

}

File: Kotlin/tollcalculator/src/main/kotlin/vehicle/Vehicle.kt
package vehicle

import java.time.LocalDateTime
import java.util.UUID

sealed class Vehicle(_lastBilledEvent: LocalDateTime, _currentDailyFee: Int) {
    var lastBilled = _lastBilledEvent
    private set
    var currentDailyFee = _currentDailyFee
    private set

    fun setLastBilled(dateTime: LocalDateTime) {
        lastBilled = dateTime
    }

    fun updateCurrentDailyFee(fee: Int) {
        currentDailyFee += fee
    }

    data class UnknownVehicle(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Car(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Motorbike(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Tractor(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Emergency(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Diplomat(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Foreign(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
    data class Military(val id: UUID = UUID.randomUUID()): Vehicle(LocalDateTime.MIN, 0)
}

